### ISSUE TITLE
Ability to link adjustments to Sale's installment contract

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#2281]
_Imported from trac ticket 2281.
Created by **sal**
Opened in trac: Apr 10, 2013, 12:24:27 PM
Last modified in trac: Sep 24, 2013, 4:15:10 PM_
***
_by sal:_
Be sure to plan for CRM/Returns associated with installment contracts. I know that debits and credits were not setup to tie back to an installment account and that needs to be fixed so maybe with returns it can be done from the start.
___
## Design notes
We should probably make this checkbox a 'pseudo' passed_attribute so it works with existing gers orders.  Likely in 'copied_passed_attributes' so rules can modify whether it should be checked, etc.?

**Comment by kentbower** [comment]
_**kbower** trac comment on Apr 10, 2013, 12:34:16 PM:_

_kbower wrote:_
We never were able to definitively say what the correct approach is for MCR/MDB tied to installment, or at least I was never made aware if that was resolved what the approach should be.
___
_sal wrote:_
The debit and credit should be able to be tied to a contract number thus it would increase or decrease the balance on the contract. It should not affect the terms of the contract or recalculate any interest, etc. I think the way GERS works when you are creating debits, credits and credit memos it asks for a contract number to associate it with which would set the AR type to "I" on the transaction.
___
_kbower wrote:_
Sal,
See our chain below.  Are you requesting that any MDB, MCR, CRM that is written against a sale that is contracted automatically be set to "I" with the contract number on the transaction?  That's kind of what it sounds like to me, though GERS asks instead. 

Does it matter whether the Sale is added to installment contract before vs. after the credit/debit is written, or should the credit/debit always go against the contract?

Does it matter whether the adjustment is credit vs. debit?  In other words, a debit is assuming more debt, should that also go against the contract?
___
_sal wrote:_
After talking with accounting the system should probably ask when creating the CRM if it needs to be tied back to an installment and give the user the opportunity to select that contract number. If the user selects a contract number then the AR type should be set to “I”. I’m always scared to say always, but I’m pretty confident that for a MDB or MCR it should always go against the contract if there is one.

If I understand your second question correctly, if at the time the credit/debit is written there is no contract on the sale that the MCR/MDB is adjusting then it would be left on the open side. Then if a contract is written then the contract would be written for the remaining balance on the order. (I would think that would be a very rare scenario since most contracts are written at the point of sale)

It doesn’t matter whether the adjustment is a credit or debit, it should go against the contract
___
_rneville wrote:_
So... I read that as:
CRM = ask
MCR/MDB = automatic
___
Yes, thanks
___
_kbower wrote:_
I don't think it makes sense to allow the user to choose the contract number.  It either applies to the sale's contract or doesn't... no point in allowing the user to mess up the contract number.  If there is some extraordinarily rare case where the contract number shouldn't be the same, it would be cleaner to do dollar transfers at that point anyway, so I don't think we will allow them to choose the contract number.

Let me state part of the point of the before vs. after question. Suppose it did occur that a credit (or other adjustment) was written (not against the contract) and then later a contract is written for the sale:  In this case, should the MCR/MDB/CRM adopt the new contract number, or should it be left alone (with no contract)?

Along that same line, I'm not sure Loft should "prompt" them -- giving them a one shot chance -- when the adjustment is written.  I'm thinking it should be a flag that they can check or uncheck, either at creation, or at a later time, that ties/removes the adjustment to the contract.  Would that make more sense?

Finally, if such an adjustment is written against the contract, must anything be recorded on the SO header (of the MCR/MDB/CRM), or is it enough to just modify the AR_TRN?
___
Returns ticket: #1037

**Comment by kentbower** [comment]
_**kbower** trac comment on Apr 10, 2013, 3:55:22 PM:_

_Sal wrote:_
1-      Agree, better to just say it does or doesn’t apply.
 
2-      Yes it should adopt the new contract number
 
3-      Yes a flag sounds good
 
4-      No just modifying the AR_TRN should be good

**Comment by kentbower** [comment]
_**kbower** trac comment on Aug 27, 2013, 4:26:07 PM:_

(Changeset [[4490]](../commit/50c5133bf9eadd023c53b720c727585d8dbc0317))
* rewrite of client-side calculated financial fields to be server-side in preparation re [#2420](2420), #2281 and #2224

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 19, 2013, 9:39:52 AM:_

From: Kent Bower [mailto:kb@retailarchitects.com]
Sent: Tuesday, September 17, 2013 11:16 AM
To: SAL
Cc: Rob Neville
Subject: Re: Your help needed, please, with financing issues


Sal, an example of these would be very helpful in order to finish implementing your request to be able to tie Credits/Returns/Debits to a contract.  Can you point be to an example where a Credit, Return or Debit is tied to a contract?  I had assumed this would mean the ORIG_FI_AMT would be > 0, but I find very very few cases of this in your database.

Thanks!
___
Take a look at contract 0214305N44. I just created credit memo 0917305EPPU tied to this installment on TRAIN.


___
Hmmm... I had expected that the CRM's ORIG_FI_AMT would have the amount there, like a SAL does when it is to be contracted....

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 19, 2013, 12:19:22 PM:_

(Changeset [[4556]](../commit/c45337f43b75335f6ea6c007e19a86e23fbf289e))
* fixes [#2499](2499), order transactions that have been archived should still be loadable, just not changable (see #2281) #all

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 23, 2013, 9:54:29 AM:_

_From Sal_
Just discovered an issue regarding installments on this. It has to do with add-on contract or refinanced contracts. See order 0923320CM98. The contract 0923320M99 that was written to pay for this order has since had an add-on done which paid off this contract. Then when I did an MCR it tied the credit to the original contract 0923320M99 from the order which is already paid off because of the add-on. In reality that credit should be applied to the new contract 0923320N12.

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 23, 2013, 12:58:23 PM:_

Rob, can you please write a Sale class method that returns the current open contractid that should be used for self?

**Comment by kentbower** [comment]
_**rneville** trac comment on Sep 23, 2013, 2:33:35 PM:_

(Changeset [[4569]](../commit/365087b87f0f0a15267b2332f0dc65e62a6ceb2c))
* references #2281 (Ability to link adjustments to Sale's installment contract) #isf

**Comment by kentbower** [comment]
_**rneville** trac comment on Sep 23, 2013, 2:34:57 PM:_

Call get_current_contractid(), you'll either get the "real" contractid to use or None.

Rob Neville

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 24, 2013, 9:57:05 AM:_

From: Kent Bower [mailto:kb@retailarchitects.com]
Sent: Tuesday, September 24, 2013 8:36 AM
To: sal@ivansmith.com
Subject: Re: #2281: Ability to link adjustments to Sale's installment contract

Just verifying: In other words, it is ok for a paid off contract to be adjusted by MCR/CRM/MDBs, correct?
___
_Sal:_ Yes, that would happen only if there were no other open contracts to apply the adjustment to. What would most likely happen in that case though is the user would uncheck the box so the adjustment wouldn’t be linked to the contract.
___
_Kent:_ That's partly why I ask... should that be an error on the check box, or would you allow it?  It sounds like you would allow it. 
___
On 9/24/2013 9:55 AM, sal@ivansmith.com wrote:
>
> The more I think about that situation, it should be an error on the check box.

___
_Kent:_ Oh, even for the case of MDB then?
___
_Sal:_ Yea

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 24, 2013, 10:00:20 AM:_

_Also from Sal:_
you should check and see if the original contract is paid off because although unlikely the customer could have multiple contracts. If it’s not paid off then the adjustment should be tied to that original contract even if there is a more recent add on because that most recent add on has no relationship to the original since the original is not paid off.
___
_In other words, I interpret this to mean we also need to check the original Sale's contract to make sure it is paid off before looking for an add on contract_

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 24, 2013, 11:32:48 AM:_

Yes, this stuff gets complicated! I guess that’s why GERS just leaves it up to the user to enter the contract number on the adjustment
___
From: Kent Bower [mailto:kb@retailarchitects.com]
Sent: Tuesday, September 24, 2013 9:37 AM
To: sal@ivansmith.com
Subject: Re: #2281: Ability to link adjustments to Sale's installment contract

Also, to verify: what if customer code changes along that path of add-ons?  I assume we should still follow that and use the new add-on?  For example, if this were a CRM, should the CRM's artran have the add-on's customerid instead of the CRM's customerid, so that it matches the add-on contract?

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 24, 2013, 12:46:24 PM:_

Sal,
    There are a handful of cases where an addon is not voided even though the original contract is voided.  Is that a corrupted condition and can we assume that no addon will exist underneath a voided contract?

Kent

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 24, 2013, 1:23:20 PM:_

FYI:  Rob and I discussed and this doesn't seem to make any sense so if they are adjusting an order where this has happened and try to check the "To Installment Contract" flag, then Loft will report "No open installment contract found." as an error on that flag.

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 24, 2013, 4:15:10 PM:_

(Changeset [[4579]](../commit/a721ffddc356670f651f8815bc8843e1f8dff5e5))
* Add ability to link Returns/Credits/Debits to Sale's third party financing #all
 * closes #2281, Ability to link adjustments to Sale's installment contract #isf
 * fixes [#2420](2420), Balance Due display (vs. Balance) work out amounts correctly on Financed orders with Refunds #all
 * closes #2224, Credits and Returns adjusting a closed financed order should credit (transfer) liability to third party finance company #all

**Comment by kentbower** [comment]
_**kbower** trac status change on Sep 24, 2013, 4:15:10 PM:_ **closed**
